### ISSUE TITLE
Redirect (via HTTP Location header) requires Back command twice to return from

### ISSUE THREAD

**Issue by jstetten** [RequestPolicyContinued/requestpolicy#780]
__Steps to reproduce:__

Trigger the redirection [notification bar](https://cloud.githubusercontent.com/assets/4286829/12374787/4daeb440-bca7-11e5-9649-0cf80b780d80.png) by following a link for which no rules exist. A common example would be basically any search results page. Click `Allow` in the bar. After reading, invoke the browser `Back` command to return.

__What happens?__

A reloading of the same page. One needs to press `Back` twice to get back to where one was originally. 

__What should happen?__

The browser should be smart enough to return to the search results page after one `Back`, not two. Breaking the browser's `Back` functionality probably confuses quite a few people in these situations.

**Comment by myrdd** [comment]
Sorry, I must correct myself. It's actually only the case for _JavaScript_ "redirections". (I've edited the issues's title.)

The solution could be simple: use [`Location.replace()`](https://developer.mozilla.org/en-US/docs/Web/API/Location/replace) instead of [`Location.assign()`](https://developer.mozilla.org/en-US/docs/Web/API/Location/assign). However, I'm quite sure there are cases where "replace" should _not_ be used, e.g. when a link on the page makes the redirect. I need to dig deeper.

**Comment by myrdd** [comment]
> there are cases where "replace" should not be used

Indeed: JavaScript code could call one of the location functions at any time. In general, when `Location.assign()` is called, we shouldn't be calling `Location.replace()`, since that page would be missing in the session history.

I was looking for a way to detect if "assign" or "replace" has been called. In fact, `nsDocShell::InternalLoad()` gets the `aLoadType` parameter, which again could be checked for the `LOAD_FLAGS_REPLACE_HISTORY` flag. However, afaict `InternalLoad` does _not_ pass this information to `shouldLoad`. So, unfortunately, there's no way to detect the "replace" parameter directly.

It should be possible to work around that problem. The easiest solution should be to hard-code which URLs should be replaced. For example, `https://www.google.[a-z]+/url`, or `https://r.search.yahoo.com/`. It's not a long-term solution, but should be fine for now.

@jstetten On which sites do you have this problem? Is it only search result pages?

**Comment by azrdev** [comment]
@myrdd back-and-forth navigation in `pdf.js` is also broken, even though I'm not sure if that's the same issue.
Open for example [this paper](https://pdfs.semanticscholar.org/ed76/a67d822d19c9633e113eb551c1fcbc3fa6b3.pdf) in `pdf.js` and click on some reference (like `[1]`), which will take you to the literature list at the end. Telling firefox to go back (via mouse buttons, context menu, or tool buttons) does not jump to the previous location, but triggers the *redirect notification bar* (and, upon selecting "allow", reloads the whole document, at the place in the literature list).
